californium-core/src/main/java/org/eclipse/californium/core/network/BaseMatcher.java

```diff
@@ -290,6 +290,34 @@ public void cancelObserve(Token token) {
 		}
 	}
 
+	protected void addCleaner(final Exchange exchange, final Message message) {
+		message.addMessageObserver(new MessageObserverAdapter() {
+
+			@Override
+			public void onCancel() {
+				complete("canceled");
+			}
+
+			@Override
+			public void failed() {
+				complete("failed");
+			}
+
+			private void complete(final String action) {
+				if (exchange.executeComplete()) {
+					if (exchange.isOfLocalOrigin()) {
+						LOG.debug("{}, {} request [MID={}, {}]", exchange, exchange.getRequest().getMID(),
```

Comment:
Uups, I forgot to add the `action` as first parameter to the log message in my commit 28e20c025f2328be491b25c2d777192106c97533.

Reply:
Uups, I didn't read this part of the code, too many trust on you ;-)

Did I change for something like : 

```java
LOG.debug("{}, {} request [MID={}, Token={}]", exchange, action, exchange.getRequest().getMID(),
								exchange.getRequest().getToken());
```

Why should we check isOfLocalOrigin ? Why sometime we are using response MID/token and sometime request ones ?

Reply:
The local origin is for requests, the remote origin for CON responses (at least in my PR).